### PR TITLE
feat: change the flow of data preprocess and avoid bug in remove columns

### PR DESCRIPTION
**Pull request by pphuc25**
# What does this PR do?
I change the data flow of prepare_dataset function, make a case to avoid remove `speech` columns

While examining the 'wav2vec2' workflow, I noticed that the `prepare_dataset` function typically takes the path of audio files and converts them into audio arrays. However, I believe this approach may not be ideal for several reasons:
- Not all data entries contain a `path` column, or the `path` column may not always be correctly populated (e.g., in the case of 'vivos' data). When attempting to use this code with such data, errors can occur.
- This process is somewhat redundant, especially in cases like 'common voice' datasets, where we already have the audio data stored in the `audio` column. In these instances, it would be more efficient to directly pass the audio array to the `speech` column.

To address these issues, I've adjusted the data flow to accept the audio file path as an input column, ensuring that the sampling rate matches the feature extractor's requirements. Additionally, I've created a list of columns to exclude during data processing to prevent inadvertently removing the 'speech' column."

I would like cc @sanchit-gandhi to review my PR.

### CODE REVIEW

File: examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py [hunk 1]
```diff
@@ -102,9 +101,9 @@ class DataTrainingArguments:
             )
         },
     )
-    speech_file_column: Optional[str] = field(
```
Comment (sanchit-gandhi) [inline examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py:105]: Let's keep this argument name the same so as not to introduce a breaking change

File: examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py [hunk 2]
```diff
@@ -329,14 +328,20 @@ def main():
...
         return batch
 
+    remove_columns_values = datasets["train"].column_names.copy()
+    if "speech" in set(remove_columns_values):
+        remove_columns_values.remove("speech")
+
     # load audio files into numpy arrays
     vectorized_datasets = datasets.map(
-        prepare_dataset, num_proc=data_args.preprocessing_num_workers, remove_columns=datasets["train"].column_names
+        prepare_dataset, num_proc=data_args.preprocessing_num_workers, remove_columns=remove_columns_values
```
Comment (sanchit-gandhi) [inline examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py:344]: Why change this? It was fine before no?

File: examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py [hunk 3]
```diff
@@ -329,14 +328,20 @@ def main():
         model_args.model_name_or_path, cache_dir=model_args.cache_dir, do_normalize=True
     )
 
+    # check that all files have the correct sampling rate
+    datasets = datasets.cast_column(data_args.audio_file_column, Audio(sampling_rate=feature_extractor.sampling_rate))
+
     def prepare_dataset(batch):
-        # check that all files have the correct sampling rate
-        batch["speech"], _ = librosa.load(batch[data_args.speech_file_column], sr=feature_extractor.sampling_rate)
+        batch["speech"] = batch[data_args.audio_file_column]["array"]
```
Comment (sanchit-gandhi) [inline examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py:335]: Fine since the latest version of datasets uses `librosa` to load audio files - can you make clear on the README that users should use the latest version of Datasets (v 2.14.5 or higher)? https://github.com/huggingface/transformers/tree/main/examples/research_projects/jax-projects/wav2vec2#wav2vec2-contrastive-loss-pretraining-examples

File: src/transformers/models/persimmon/modeling_persimmon.py [hunk 4]
```diff
@@ -72,7 +72,7 @@ def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int]
 
 
 # Copied from transformers.models.llama.modeling_llama.LlamaRotaryEmbedding with Llama->Persimmon
-class PersimmonRotaryEmbedding(torch.nn.Module):
+class PersimmonRotaryEmbedding(nn.Module):
```
Comment (sanchit-gandhi) [inline src/transformers/models/persimmon/modeling_persimmon.py:75]: If you rebase onto main you'll get these changes automatically

**Review by sanchit-gandhi — COMMENTED** [review]
Hey @pphuc25 - the Flax Wav2Vec2 pre-training script is under 'research projects' because it's still very much a WIP, and currently does not have correctness (see https://github.com/huggingface/transformers/issues/19588). Thus, it's not really actively maintained, and so is not a fruitful place to make new contributions. I would encourage you to either develop on-top of the existing script and publish it standalone, or pick-up the work started in #19588 to try and get equivalence with PyTorch before adding new functionality!

File: examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py (hunk 2 above)
Comment (pphuc25) [reply examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py:344]: in function preprocess, the column will be assigned to colum name `speech`, I think maybe the bug can occur seen some data have the column `speech` will automatic remove, this is not a rare case seen something I name my audio column name as `speech`

File: examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py (hunk 3 above)
Comment (pphuc25) [reply examples/research_projects/jax-projects/wav2vec2/run_wav2vec2_pretrain_flax.py]: I have documented in readme for lastest commit

**Comment by sanchit-gandhi** [comment]
Hey @pphuc25! Thanks for your enthusiasm here! As mentioned [previously](https://github.com/huggingface/transformers/pull/26516#pullrequestreview-1653115447), this examples script is not the best place to make performance optimisations, since it's currently a WIP script (or more truthfully, a 'broken' script). If you're interested in making a contribution for Flax Wav2Vec2 pre-training, I would encourage you to take a look at the issue #19588, which endeavours to correct this script by obtaining equivalence with PyTorch. We should fix this script first before making performance optimisations, like the ones proposed in this PR. Thanks for your understanding.
